### PR TITLE
Progress logic integration

### PR DESCRIPTION
**Pull request by szabinst**
refs: MBL-16840
affects: Student
release note: none
test plan: See more below

### This PR integrates the [Progress reporting logic](https://github.com/instructure/canvas-ios/pull/2481) and contains the following changes: 
- Extracted CourseSyncEntry related entities to their own files
- Reworked CourseSyncEntry’s index based selection logic look up to be a custom id based one 
- Refactored `CourseSyncInteractor` for better readability 
- Added a clean up method to `CourseSyncProgressWriterInteractor` so file progress is reseted when a new download starts
- Added `DeleteLocalUseCase`
- Added dedicated `LocalUseCase`, `DeleteLocalUseCase` class definitions for progress related logic
- Added File size to `CourseSyncSelection` and `CourseSyncProgress` screens  
- Renamed `CourseSyncFileFolderInteractor` to `CourseSyncEntryComposerInteractor` to better express its responsibility
- Added the `getEntitiesFromDatabase` function to `ReactiveStore` so we can directly read from `Core Data`   
-  Renamed `XCTAssertSingleOutputEquals` to `XCTAssertCompletableSingleOutputEquals` and added a new `XCTAssertSingleOutputEquals` implementation so we can assert on non-completable publishers as well

### What's not part of this PR 
- Progress UI opens from a temporary kebab menu item and not from the Dashboard Notification.
- Cancellation.
- Retry.

## Testing 
0. You can use my user to test lengthy downloads as I have plenty of uploaded files 
1. Select kebab menu on the Dashboard. Select `Manage Offline Content`. 
2. Select kebab menu on the Dashboard. Select `Offline Content Downloads`. 
3. You should only see courses/tabs/files that were previously selected. 
4. You should see them progressing. 

## Screenshots

<table>
<tr><th>In Progress - Collapsed</th><th>In Progress - Expanded</th></tr>
<tr>
<td></td>
<td></td>
</tr>
<tr><th>Error state</th><th>CourseSyncSelector file size</th></tr>
<tr>
<td></td>
<td></td>
</tr>
</table>

## Checklist

- [x] A11y checked
- [x] Tested on phone
- [ ] Tested on tablet
- [ ] Tested in dark mode
- [x] Tested in light mode
- [ ] Approve from product

### CODE REVIEW

**Comment by inst-danger** [comment]
<details><summary>Student Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This pull request will not generate a release note.</td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>One or more files are below the minimum test coverage 50%</td>
    </tr>
  </tbody>
</table>


#### Affected Apps: Student

[MBL-16840](https://instructure.atlassian.net/browse/MBL-16840)

Coverage | New % | Master % | Delta
-------- | ----- | -------- | -----
**Canvas iOS** | 92.42% | 92.4% | 0.02%
Core/Core/CourseSync/CourseSyncProgress/Model/UseCase/DeleteCourseSyncFileProgressUseCase.swift | 0% | -- | --
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 231ad6ebc3541eb89a9f42cccd12c51bd2549c44
</p>

**Review by vargaat — CHANGES_REQUESTED** [review]
I don't know if we want to amend this now but the error banner is different than we have in Figma.

File: Core/Core/CourseSync/Common/Models/Entities/CourseSyncEntry+Array.swift [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    var totalDownloadedSize: Int {
+        reduce(0) { partialResult, entry in
+            partialResult + entry.totalDownloadedSize
+        }
+    }
+
+    var progress: Float {
+        Float(totalDownloadedSize) / Float(totalSelectedSize)
```
Comment (idevainstr) [inline Core/Core/CourseSync/Common/Models/Entities/CourseSyncEntry+Array.swift:35]: Could we have a zero check here and at the other instances of divisions just to be safe?

**Review by idevainstr — CHANGES_REQUESTED** [review]

File: Core/CoreTests/CourseSync/CourseSyncProgress/Model/CourseSyncProgressObserverInteractorLiveTests.swift [hunk 2]
```diff
@@ -209,70 +209,66 @@ class CourseSyncProgressObserverInteractorLiveTests: CoreTestCase {
...
+
+        let expectation = expectation(description: "Publisher sends value")
+        let subscription = testee.observeFileProgress()
+            .dropFirst()
+            .sink { state in
+                if case let .data(list) = state {
+                    print(list[0].bytesDownloaded)
+                    print(list[0].bytesToDownload)
+                    print(list[0].progress)
+
```
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/CoreTests/CourseSync/CourseSyncProgress/Model/CourseSyncProgressObserverInteractorLiveTests.swift:263]: Fixed.

File: Core/CoreTests/CourseSync/CourseSyncProgress/ViewModel/CourseSyncProgressInfoViewModelTests.swift [hunk 3]
```diff
@@ -16,16 +16,73 @@
...
 import CombineExt
+@testable import Core
+import XCTest
+
+class CourseSyncProgressInfoViewModelTests: CoreTestCase {
+    private var courseSyncProgressInteractorMock: MockCourseSyncProgressInteractor!
+
+    override func setUp() {
+        courseSyncProgressInteractorMock = MockCourseSyncProgressInteractor()
+        super.setUp()
```
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/CoreTests/CourseSync/CourseSyncProgress/ViewModel/CourseSyncProgressInfoViewModelTests.swift:29]: Fixed.

File: Core/Core/CourseSync/CourseSyncDownloader/Model/CourseSyncProgressWriterInteractor.swift [hunk 4]
```diff
@@ -16,34 +16,44 @@
...
             try? context.save()
         }
     }
 
-    func saveEntryProgress(id: String, selection: CourseEntrySelection, state: CourseSyncEntry.State) {
+    public func cleanUpPreviousFileProgress() {
+        context.performAndWait {
+            context.delete(context.fetch(scope: .all) as [CourseSyncFileProgress])
+            try? context.save()
+        }
```
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/Core/CourseSync/CourseSyncDownloader/Model/CourseSyncProgressWriterInteractor.swift:53]: Fixed.

File: Core/Core/SwiftUIViews/ListCellView/ListCellView.swift [hunk 5]
```diff
@@ -133,10 +165,12 @@ struct ListCellView: View {
 
     @ViewBuilder
     private var errorText: some View {
-        Text(viewModel.error ?? "")
-            .lineLimit(1)
-            .foregroundColor(.textDanger)
-            .font(.regular14)
+        if case .error(let error) = viewModel.state {
+            Text(error ?? "")
```
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/Core/SwiftUIViews/ListCellView/ListCellView.swift:169]: Fixed.

File: Core/Core/CourseSync/Common/Models/Entities/CourseSyncEntry+Array.swift (hunk 1 above)
Comment (szabinst) [reply Core/Core/CourseSync/Common/Models/Entities/CourseSyncEntry+Array.swift:35]: Fixed.

File: Core/Core/SwiftUIViews/ListCellView/ListCellViewModel.swift [hunk 6]
```diff
@@ -128,13 +126,13 @@ class ListCellViewModel: ObservableObject {
...
-                progressText = NSLocalizedString("Downloading", bundle: .core, comment: "")
-            }
-        }
+//        if let progress = progress, error == nil {
+//            if progress == 1 {
+//                progressText = NSLocalizedString("Download complete", bundle: .core, comment: "")
+//            } else {
+//                progressText = NSLocalizedString("Downloading", bundle: .core, comment: "")
+//            }
+//        }
```
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/Core/SwiftUIViews/ListCellView/ListCellViewModel.swift:135]: Fixed.

File: Core/Core/SwiftUIViews/ListCellView/ListCellViewModel.swift [hunk 7]
```diff
@@ -103,9 +101,9 @@ class ListCellViewModel: ObservableObject {
 
     var accessibilityText: String {
         var titleText = title + (subtitle ?? "")
-        if let error = error {
-            titleText.append("," + error)
-        }
+//        if let error = error {
+//            titleText.append("," + error)
+//        }
```
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/Core/SwiftUIViews/ListCellView/ListCellViewModel.swift:106]: Fixed.

**Review by balintbartok — DISMISSED** [review]

**Review by idevainstr — DISMISSED** [review]

**Review by idevainstr — APPROVED** [review]
QA +1

**Review by idevainstr — APPROVED** [review]

**Review by vargaat — APPROVED** [review]
QA+1

**Pull request merged**